pkg/runtime/wfengine/activity.go

```diff
@@ -0,0 +1,298 @@
+			wfLogger.Warnf("%s: execution of '%s' timed-out and will be retried later", actorID, reminderName)
+
+			// Returning nil signals that we want the execution to be retried in the next period interval
+			return nil
+		} else if _, ok := err.(recoverableError); ok {
+			wfLogger.Warnf("%s: execution failed with a recoverable error and will be retried later: %v", actorID, err)
+
+			// Returning nil signals that we want the execution to be retried in the next period interval
+			return nil
+		} else {
+			wfLogger.Errorf("%s: execution failed with a non-recoverable error: %v", actorID, err)
+			// TODO: Reply with a failure - this requires support from durabletask-go to produce TaskFailure results
+		}
+	}
+
+	// TODO: Purge actor state based on some data retention policy
+
+	// We delete the reminder on success and on non-recoverable errors.
+	return actors.ErrReminderCanceled
+}
+
+func (a *activityActor) executeActivity(ctx context.Context, actorID string, name string, eventPayload []byte) error {
+	taskEvent, err := backend.UnmarshalHistoryEvent(eventPayload)
+	if err != nil {
+		return err
+	}
+
+	endIndex := strings.LastIndex(actorID, "#")
+	if endIndex < 0 {
+		return fmt.Errorf("invalid activity actor ID: %s", actorID)
+	}
+	workflowID := actorID[0:endIndex]
+
+	wi := &backend.ActivityWorkItem{
+		SequenceNumber: int64(taskEvent.EventId),
+		InstanceID:     api.InstanceID(workflowID),
+		NewEvent:       taskEvent,
+		Properties:     make(map[string]interface{}),
+	}
+
+	// Executing activity code is a one-way operation. We must wait for the app code to report its completion, which
+	// will trigger this callback channel.
+	// TODO: Need to come up with a design for timeouts. Some activities may need to run for hours but we also need
+	//       to handle the case where the app crashes and never responds to the workflow. It may be necessary to
+	//       introduce some kind of heartbeat protocol to help identify such cases.
+	callback := make(chan bool)
+	wi.Properties[CallbackChannelProperty] = callback
+	if err = a.scheduler.ScheduleActivity(ctx, wi); err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			return newRecoverableError(errors.New(
+				"timed-out trying to schedule an activity execution - this can happen if too many activities are running in parallel or if the workflow engine isn't running"))
+		}
+		return newRecoverableError(fmt.Errorf("failed to schedule an activity execution: %w", err))
+	}
+
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(10 * time.Minute):
+			if deadline, ok := ctx.Deadline(); ok {
+				wfLogger.Warnf("%s: '%s' is still running - will keep waiting until %v", actorID, name, deadline)
+			} else {
+				wfLogger.Warnf("%s: '%s' is still running - will keep waiting indefinitely", actorID, name)
+			}
+		case completed := <-callback:
+			if completed {
+				break loop
+			} else {
+				return newRecoverableError(errExecutionAborted)
+			}
```

Comment:
nit: in case of the callback channel being closed this will return `executionAborted`, is this expected?
```suggestion
		case completed, ok := <-callback:
			if completed || !ok {
				break loop
			} else {
				return newRecoverableError(errExecutionAborted)
			}
```

Reply:
I think so. If the channel is closed I assume that means something is shutting down, in which case aborting the execution is the right thing to do (it will be retried later). We definitely don't want to continue execution because then the workflow would receive an undefined response from the activity it scheduled. Activities must always be allowed to run until completion or until an explicit failure.